File: 05_Bagels/python/bagels.py
######################################################################
#
# Bagels
#
# From: BASIC Computer Games (1978)
#       Edited by David H. Ahl
#
# "In this game, the computer picks a 3-digit secret number using
#  the digits 0 to 9 and you attempt to guess what it is.  You are
#  allowed up to twenty guesses.  No digit is repeated.  After
#  each guess the computer will give you clues about your guess
#  as follows:
#
#  PICO     One digit is correct, but in the wrong place
#  FERMI    One digit is in the correct place
#  BAGELS   No digit is correct
#
# "You will learn to draw inferences from the clues and, with
#  practice, you'll learn to improve your score.  There are several
#  good strategies for playing Bagels.  After you have found a good
#  strategy, see if you can improve it.  Or try a different strategy
#  altogether and see if it is any better.  While the program allows
#  up to twenty guesses, if you use a good strategy it should not
#  take more than eight guesses to get any number.
#
# "The original authors of this program are D. Resek and P. Rowe of
#  the Lawrence Hall of Science, Berkeley, California."
#
#
# Python port by Jeff Jetton, 2019
#
######################################################################


import random

MAX_GUESSES = 20



def print_rules():
    print("\nI am thinking of a three-digit number.  Try to guess")
    print("my number and I will give you clues as follows:")
    print("   PICO   - One digit correct but in the wrong position")
    print("   FERMI  - One digit correct and in the right position")
    print("   BAGELS - No digits correct")



def pick_number():
    # Note that this returns a list of individual digits
    # as separate strings, not a single integer or string
    numbers = [i for i in range(10)]
    random.shuffle(numbers)
    num = numbers[0:3]
    num = [str(i) for i in num] 
    return num



def get_valid_guess():
    valid = False
    while not valid:
        guess = input(f"Guess # {guesses}     ? ")
        guess = guess.strip()
        # Guess must be three characters
        if len(guess) == 3:
            # And they should be numbers
            if guess.isnumeric():
                # And the numbers must be unique
                if len(set(guess)) == 3:
                    valid = True
                else:
                    print("Oh, I forgot to tell you that " +
                          "the number I have in mind")
                    print("has no two digits the same.")
            else:
                print("What?")
        else:
            print("Try guessing a three-digit number.")

    return guess



def build_result_string(num, guess):

    result = ""

    # Correct digits in wrong place
    for i in range(2):
        if num[i] == guess[i+1]:
            result += "PICO "
        if num[i+1] == guess[i]:
            result += "PICO "
    if num[0] == guess[2]:
        result += "PICO "
    if num[2] == guess[0]:
        result += "PICO "

    # Correct digits in right place
    for i in range(3):
        if num[i] == guess[i]:
            result += "FERMI "

    # Nothing right?
    if result == "":
        result = "BAGELS"

    return result




######################################################################


# Intro text
print("\n                Bagels")
print("Creative Computing  Morristown, New Jersey")
print("\n\n")

# Anything other than N* will show the rules
response = input("Would you like the rules (Yes or No)? ")
if len(response) > 0:
    if response.upper()[0] != 'N':
        print_rules()
else:
    print_rules()

games_won = 0
still_running = True
while still_running:

    # New round
    num = pick_number()
    num_str = ''.join(num)
    guesses = 1

    print("\nO.K.  I have a number in mind.")
    guessing = True
    while guessing:

        guess = get_valid_guess()

        if guess == num_str:
            print("You got it!!!\n")
            games_won += 1
            guessing = False
        else: 
            print(build_result_string(num, guess))
            guesses += 1
            if guesses > MAX_GUESSES:
                print("Oh well")
                print(f"That's {MAX_GUESSES} guesses.  " +
                      "My number was " + num_str)
                guessing = False   


    valid_response = False
    while not valid_response:
        response = input("Play again (Yes or No)? ")
        if len(response) > 0:
            valid_response = True
            if response.upper()[0] != "Y":
                still_running = False


if games_won > 0:
    print(f"\nA {games_won} point Bagels buff!!")

print("Hope you had fun.  Bye.\n")



######################################################################
#
# Porting Notes
#
#   The original program did an unusually good job of validating the
#   player's input (compared to many of the other programs in the
#   book). Those checks and responses have been exactly reproduced.
#   
#
# Ideas for Modifications
#
#   It should probably mention that there's a maximum of MAX_NUM
#   guesses in the instructions somewhere, shouldn't it?
#
#   Could this program be written to use anywhere from, say 2 to 6
#   digits in the number to guess? How would this change the routine
#   that creates the "result" string?
#
######################################################################


            
        
    

    

File: 05_Bagels/javascript/bagels.js
// BAGELS
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

print(tab(33) + "BAGELS\n");
print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");

// *** Bagles number guessing game
// *** Original source unknown but suspected to be
// *** Lawrence Hall of Science, U.C. Berkeley

a1 = [0,0,0,0];
a = [0,0,0,0];
b = [0,0,0,0];

y = 0;
t = 255;

print("\n");
print("\n");
print("\n");

// Main program
async function main()
{
    while (1) {
        print("WOULD YOU LIKE THE RULES (YES OR NO)");
        str = await input();
        if (str.substr(0, 1) != "N") {
            print("\n");
            print("I AM THINKING OF A THREE-DIGIT NUMBER.  TRY TO GUESS\n");
            print("MY NUMBER AND I WILL GIVE YOU CLUES AS FOLLOWS:\n");
            print("   PICO   - ONE DIGIT CORRECT BUT IN THE WRONG POSITION\n");
            print("   FERMI  - ONE DIGIT CORRECT AND IN THE RIGHT POSITION\n");
            print("   BAGELS - NO DIGITS CORRECT\n");
        }
        for (i = 1; i <= 3; i++) {
            do {
                a[i] = Math.floor(Math.random() * 10);
                for (j = i - 1; j >= 1; j--) {
                    if (a[i] == a[j])
                        break;
                }
            } while (j >= 1) ;
        }
        print("\n");
        print("O.K.  I HAVE A NUMBER IN MIND.\n");
        for (i = 1; i <= 20; i++) {
            while (1) {
                print("GUESS #" + i);
                str = await input();
                if (str.length != 3) {
                    print("TRY GUESSING A THREE-DIGIT NUMBER.\n");
                    continue;
                }
                for (z = 1; z <= 3; z++)
                    a1[z] = str.charCodeAt(z - 1);
                for (j = 1; j <= 3; j++) {
                    if (a1[j] < 48 || a1[j] > 57)
                        break;
                    b[j] = a1[j] - 48;
                }
                if (j <= 3) {
                    print("WHAT?");
                    continue;
                }
                if (b[1] == b[2] || b[2] == b[3] || b[3] == b[1]) {
                    print("OH, I FORGOT TO TELL YOU THAT THE NUMBER I HAVE IN MIND\n");
                    print("HAS NO TWO DIGITS THE SAME.\n");
                    continue;
                }
                break;
            }
            c = 0;
            d = 0;
            for (j = 1; j <= 2; j++) {
                if (a[j] == b[j + 1])
                    c++;
                if (a[j + 1] == b[j])
                    c++;
            }
            if (a[1] == b[3])
                c++;
            if (a[3] == b[1])
                c++;
            for (j = 1; j <= 3; j++) {
                if (a[j] == b[j])
                    d++;
            }
            if (d == 3)
                break;
            for (j = 0; j < c; j++)
                print("PICO ");
            for (j = 0; j < d; j++)
                print("FERMI ");
            if (c + d == 0)
                print("BAGELS");
            print("\n");
        }
        if (i <= 20) {
            print("YOU GOT IT!!!\n");
            print("\n");
        } else {
            print("OH WELL.\n");
            print("THAT'S A TWENTY GUESS.  MY NUMBER WAS " + a[1] + a[2] + a[3]);
        }
        y++;
        print("PLAY AGAIN (YES OR NO)");
        str = await input();
        if (str.substr(0, 1) != "Y")
            break;
    }
    if (y == 0)
        print("HOPE YOU HAD FUN.  BYE.\n");
    else
        print("\nA " + y + " POINT BAGELS BUFF!!\n");

}

main();
